File: apps/api-rs/src/state.rs
use std::sync::Arc;

use anyhow::Result;
use sqlx::PgPool;

use crate::agent::{
    AgentKernel, AgentModel, AgentToolRegistry, BuiltinPromptRegistry, ClarificationTool,
    GroundedAnswerFinalizer, KnowledgeSearchTool, LlmClaimVerifier, StructuralClaimVerifier,
};
use crate::config::AppConfig;
use crate::llm::openai::{OpenAiClient, OpenAiClientConfig};
use crate::rag::{
    EmbeddingClientConfig, EsRetriever, HttpReranker, RerankProvider, SimpleContextAssembler,
};
use crate::repositories::{InMemoryConversationRepository, SqlxConversationRepository};
use crate::storage::{build_storage, ObjectStorage};
use tracing::warn;

#[derive(Clone)]
pub struct AppState {
    pub config: AppConfig,
    pub repository: Arc<dyn crate::repositories::ConversationRepository>,
    pub agent_kernel: AgentKernel,
    pub db_pool: Option<PgPool>,
    pub redis_client: Option<redis::Client>,
    pub storage: Arc<dyn ObjectStorage>,
}

pub async fn build_state(config: AppConfig) -> Result<AppState> {
    let (repository, db_pool): (
        Arc<dyn crate::repositories::ConversationRepository>,
        Option<PgPool>,
    ) = if let Some(ref url) = config.database_url {
        let pool = sqlx::PgPool::connect(url).await?;
        crate::auth::seed_identity(&pool, &config).await?;
        recover_interrupted_agent_runs(&pool).await?;
        if let Err(err) = crate::api::documents::recover_interrupted_document_jobs(&pool).await {
            warn!(error = %err, "failed to recover interrupted document jobs");
        }
        (
            Arc::new(SqlxConversationRepository::new(pool.clone())),
            Some(pool),
        )
    } else {
        (Arc::new(InMemoryConversationRepository::new()), None)
    };

    let redis_client = if let Some(ref url) = config.redis_url {
        Some(redis::Client::open(url.as_str())?)
    } else {
        None
    };

    if !config.rag.generation.use_real_llm {
        anyhow::bail!(
            "DocuMind Agent requires USE_REAL_LLM=true; rule-based answer fallback was removed"
        );
    }
    let generation_client = Arc::new(OpenAiClient::new(OpenAiClientConfig {
        base_url: config.rag.generation.base_url.clone(),
        api_key: config.rag.generation.api_key.clone(),
        model: config.rag.generation.model.clone(),
        timeout_seconds: 120,
    })?);
    let reasoning_client = Arc::new(OpenAiClient::new(OpenAiClientConfig {
        base_url: config.rag.generation.base_url.clone(),
        api_key: config.rag.generation.api_key.clone(),
        model: config.agent.reasoning_model.clone(),
        timeout_seconds: 120,
    })?);
    let agent_model: Arc<dyn AgentModel> = generation_client;
    let verifier: Arc<dyn crate::agent::ClaimVerifier> = if config.rag.citation.verify_claims {
        Arc::new(LlmClaimVerifier::new(
            reasoning_client,
            config.agent.reasoning_model.clone(),
            config.rag.citation.verify_consensus,
        ))
    } else {
        Arc::new(StructuralClaimVerifier)
    };

    if !config.rag.embedding.enabled {
        anyhow::bail!("DocuMind Agent requires EMBED_ENABLED=true");
    }
    let es_url = config
        .elasticsearch_url
        .as_ref()
        .ok_or_else(|| anyhow::anyhow!("DocuMind Agent requires ELASTICSEARCH_URL"))?;
    let embedding_config = EmbeddingClientConfig::try_from(&config.rag.embedding)?;
    let retriever: Arc<dyn crate::rag::Retriever> = Arc::new(EsRetriever::new(
        es_url.clone(),
        config.rag.embedding.index_alias.clone(),
        embedding_config,
        config.rag.embedding.model.clone(),
    )?);

    if !config.rag.rerank.enabled {
        anyhow::bail!(
            "DocuMind Agent requires RAG_RERANK_ENABLED=true; rule-based reranking was removed"
        );
    }
    let rerank_url = config
        .rag
        .rerank
        .api_url
        .clone()
        .ok_or_else(|| anyhow::anyhow!("RAG_RERANK_API_URL is required"))?;
    let reranker_adapter = HttpReranker::new(
        rerank_url,
        config.rag.rerank.api_key.clone(),
        config.rag.rerank.model.clone(),
        RerankProvider::parse(&config.rag.rerank.provider)?,
    )?;
    reranker_adapter
        .probe()
        .await
        .map_err(|error| anyhow::anyhow!("reranker readiness probe failed: {error}"))?;
    let reranker: Arc<dyn crate::rag::Reranker> = Arc::new(reranker_adapter);

    let tools = AgentToolRegistry::new(vec![
        Arc::new(KnowledgeSearchTool::new(retriever, reranker)),
        Arc::new(ClarificationTool),
    ])?;
    let agent_kernel = AgentKernel::new(
        agent_model,
        tools,
        Arc::new(SimpleContextAssembler::new()),
        Arc::new(BuiltinPromptRegistry::new()),
        Arc::new(GroundedAnswerFinalizer::new(verifier)),
    )?;

    let storage = build_storage(&config);

    if let Some(pool) = db_pool.clone() {
        crate::rag::vector_pipeline::start_vector_worker(
            pool,
            config.rag.embedding.clone(),
            config.elasticsearch_url.clone(),
            config.rabbitmq_url.clone(),
        );
    }

    let state = AppState {
        config,
        repository,
        agent_kernel,
        db_pool,
        redis_client,
        storage,
    };
    let resumed = crate::api::documents::resume_pending_document_jobs(&state)
        .await
        .map_err(|error| anyhow::anyhow!("failed to resume document jobs: {error:?}"))?;
    if resumed > 0 {
        warn!(resumed, "resumed pending document jobs");
    }
    Ok(state)
}

async fn recover_interrupted_agent_runs(pool: &PgPool) -> Result<()> {
    let recovered = sqlx::query(
        "UPDATE conversation_messages
         SET status = 'failed',
             error_code = 'EXECUTION_INTERRUPTED',
             error_message = 'Agent execution was interrupted before completion; retry this message.',
             completed_at = NOW()
         WHERE role = 'assistant' AND status = 'answering'",
    )
    .execute(pool)
    .await?
    .rows_affected();
    if recovered > 0 {
        warn!(recovered, "recovered interrupted agent messages");
    }
    Ok(())
}

File: apps/api-bun/src/state.ts
// 移植自 apps/api-rs/src/state.rs —— 应用状态装配
import type { Sql } from 'postgres';
import type Redis from 'ioredis';
import postgres from 'postgres';
import { Redis as RedisClient } from 'ioredis';
import type { AppConfig } from './config.ts';
import type { ConversationRepository } from './repositories/types.ts';
import { InMemoryConversationRepository, SqlxConversationRepository } from './repositories/index.ts';
import type { ObjectStorage } from './storage/types.ts';
import { buildStorage } from './storage/index.ts';
import { seedIdentity } from './auth/seed.ts';
import { OpenAiClient } from './llm/openai.ts';
import {
  BuiltinPromptRegistry, GroundedAnswerFinalizer, LlmClaimVerifier, PiAgentKernel,
  StructuralClaimVerifier, buildPiStreamFn, type PiModelSettings,
} from './agent/index.ts';
import type { ContextAssembler, Retriever, Reranker } from './rag/types.ts';
import { EsRetriever } from './rag/retriever.ts';
import { HttpReranker, parseRerankProvider } from './rag/reranker.ts';
import { SimpleContextAssembler } from './rag/context.ts';
import { embeddingClientConfigFrom } from './rag/embedding.ts';
import { quickConsistency, startVectorWorker } from './rag/vector_pipeline.ts';
import type { VectorConsistencySnapshot } from './http/health.ts';
import type { ClaimVerifier } from './agent/verifier/types.ts';
import { loadSystemSettings } from './system_settings.ts';

export interface AppState {
  config: AppConfig;
  sql: Sql | null;
  redis: Redis | null;
  repository: ConversationRepository;
  agentKernel: PiAgentKernel;
  storage: ObjectStorage;
  /** 健康检查用：rag/vector_pipeline.quickConsistency */
  vectorConsistency: (() => Promise<VectorConsistencySnapshot>) | null;
  llm: {
    generationClient: OpenAiClient;
    reasoningClient: OpenAiClient;
    retriever: Retriever;
    reranker: Reranker;
    contextAssembler: ContextAssembler;
  };
}

export async function buildState(config: AppConfig): Promise<AppState> {
  let sql: Sql | null = null;
  let repository: ConversationRepository;
  if (config.databaseUrl) {
    sql = postgres(config.databaseUrl, { max: 10 });
    await seedIdentity(sql, config);
    await loadSystemSettings(sql, config);
    await recoverInterruptedAgentRuns(sql);
    try {
      const { recoverInterruptedDocumentJobs } = await import('./api/documents.ts');
      await recoverInterruptedDocumentJobs(sql);
    } catch (error) {
      console.warn(`[documind][state] failed to recover interrupted document jobs: ${(error as Error).message}`);
    }
    repository = new SqlxConversationRepository(sql);
  } else {
    repository = new InMemoryConversationRepository();
  }

  let redis: Redis | null = null;
  if (config.redisUrl) {
    redis = new RedisClient(config.redisUrl, { maxRetriesPerRequest: 2 });
  }
  if (!config.rag.generation.useRealLlm) {
    throw new Error('DocuMind Agent requires USE_REAL_LLM=true; rule-based answer fallback was removed');
  }
  const generationClient = new OpenAiClient({
    baseUrl: config.rag.generation.baseUrl,
    apiKey: config.rag.generation.apiKey,
    model: config.rag.generation.model,
    timeoutSeconds: 120,
  });
  const reasoningClient = new OpenAiClient({
    baseUrl: config.rag.generation.baseUrl,
    apiKey: config.rag.generation.apiKey,
    model: config.agent.reasoningModel,
    timeoutSeconds: 120,
  });
  if (!config.rag.embedding.enabled) {
    throw new Error('DocuMind Agent requires EMBED_ENABLED=true');
  }
  const esUrl = config.elasticsearchUrl;
  if (!esUrl) {
    throw new Error('DocuMind Agent requires ELASTICSEARCH_URL');
  }
  if (!sql) {
    throw new Error('DocuMind Agent requires DATABASE_URL');
  }
  const embeddingConfig = embeddingClientConfigFrom(config.rag.embedding);
  const retriever: Retriever = new EsRetriever(
    esUrl, config.rag.embedding.indexAlias, embeddingConfig,
    config.rag.embedding.model, sql,
  );

  if (!config.rag.rerank.enabled) {
    throw new Error('DocuMind Agent requires RAG_RERANK_ENABLED=true; rule-based reranking was removed');
  }
  const rerankUrl = config.rag.rerank.apiUrl;
  if (!rerankUrl) {
    throw new Error('RAG_RERANK_API_URL is required');
  }
  const rerankerAdapter = new HttpReranker(
    rerankUrl, config.rag.rerank.apiKey, config.rag.rerank.model,
    parseRerankProvider(config.rag.rerank.provider),
  );
  await rerankerAdapter.probe();
  const reranker: Reranker = rerankerAdapter;
  const contextAssembler: ContextAssembler = new SimpleContextAssembler();

  const verifier: ClaimVerifier = config.rag.citation.verifyClaims
    ? new LlmClaimVerifier(reasoningClient, config.agent.reasoningModel, config.rag.citation.verifyConsensus)
    : new StructuralClaimVerifier();
  const modelSettings: PiModelSettings = {
    model: config.rag.generation.model,
    baseUrl: config.rag.generation.baseUrl,
    apiKey: config.rag.generation.apiKey,
    contextWindow: config.rag.generation.contextWindow,
    maxTokens: config.rag.generation.maxOutputTokens,
    temperature: config.rag.generation.temperature,
  };
  const agentKernel = new PiAgentKernel({
    settings: modelSettings,
    streamFn: buildPiStreamFn(modelSettings),
    retriever: retriever,
    reranker: reranker,
    contextAssembler: contextAssembler,
    promptRegistry: new BuiltinPromptRegistry(),
    answerFinalizer: new GroundedAnswerFinalizer(verifier),
  });

  const storage = buildStorage(config);
  const vectorConsistency = () => quickConsistency(sql, config.rag.embedding, esUrl);
  // 与 Rust 一致：有数据库时启动后台向量 worker（DB 轮询消费，无 AMQP 队列加速）
  startVectorWorker(sql, config.rag.embedding, esUrl, config.rabbitmqUrl);

  const state: AppState = {
    config, sql, redis, repository, agentKernel, storage, vectorConsistency,
    llm: { generationClient, reasoningClient, retriever, reranker, contextAssembler },
  };
  // 与 Rust 一致：恢复上次中断遗留的待处理解析任务
  try {
    const { resumePendingDocumentJobs } = await import('./api/documents.ts');
    const resumed = await resumePendingDocumentJobs(state);
    if (resumed > 0) {
      console.warn(`[documind][state] resumed ${resumed} pending document jobs`);
    }
  } catch (error) {
    throw new Error(`failed to resume document jobs: ${(error as Error).message}`);
  }
  return state;
}

async function recoverInterruptedAgentRuns(sql: Sql): Promise<void> {
  const result = await sql.unsafe(`
    UPDATE conversation_messages
    SET status = 'failed',
        error_code = 'EXECUTION_INTERRUPTED',
        error_message = 'Agent execution was interrupted before completion; retry this message.',
        completed_at = NOW()
    WHERE role = 'assistant' AND status = 'answering'
  `);
  if (result.count > 0) {
    console.warn(`[documind][state] recovered ${result.count} interrupted agent messages`);
  }
}
